#################################################################
## Effect of read extension on peaks for short read sequencing ##
#################################################################

library(shiny)
library(ggplot2)
library(ggbio)
library(GenomicRanges)

# Data from chipseq::cstest$ctcf in a window chr10:80.75-80.753Mb
dat <- GRanges(seqnames="chr10",IRanges(c(80750267,80750463,80750504,80750520,80750526,80750543,80750555,80750557,80750563,80750566,80750581,80750582,80750583,80750584,80750587,80750591,80750593,80750596,80750600,80750616,80750623,80750626,80750628,80750630,80750634,80750636,80750642,80750645,80750646,80750647,80750649,80750651,80750652,80750653,80750654,80750664,80750680,80750682,80750683,80750687,80750758,80750872,80751114,80751382,80751395,80751410,80751434,80751451,80751469,80751488,80751500,80751501,80751503,80751504,80751509,80751520,80751525,80751555,80751565,80751636,80752083,80752105,80752114,80752133,80752137,80752185,80752194,80752200,80752202,80752217,80752225,80752247,80752618,80750107,80750643,80750658,80750659,80750660,80750666,80750667,80750668,80750670,80750675,80750676,80750677,80750678,80750688,80750694,80750696,80750701,80750705,80750706,80750707,80750710,80750717,80750719,80750720,80750721,80750734,80750747,80750755,80750764,80750769,80750771,80750811,80750851,80750857,80750936,80751344,80751515,80751536,80751547,80751548,80751568,80751581,80751587,80751596,80751600,80751629,80751630,80751637,80751646,80751651,80751868,80751931,80752178,80752198,80752252,80752265,80752287,80752323,80752330,80752556),c(80750290,80750486,80750527,80750543,80750549,80750566,80750578,80750580,80750586,80750589,80750604,80750605,80750606,80750607,80750610,80750614,80750616,80750619,80750623,80750639,80750646,80750649,80750651,80750653,80750657,80750659,80750665,80750668,80750669,80750670,80750672,80750674,80750675,80750676,80750677,80750687,80750703,80750705,80750706,80750710,80750781,80750895,80751137,80751405,80751418,80751433,80751457,80751474,80751492,80751511,80751523,80751524,80751526,80751527,80751532,80751543,80751548,80751578,80751588,80751659,80752106,80752128,80752137,80752156,80752160,80752208,80752217,80752223,80752225,80752240,80752248,80752270,80752641,80750130,80750666,80750681,80750682,80750683,80750689,80750690,80750691,80750693,80750698,80750699,80750700,80750701,80750711,80750717,80750719,80750724,80750728,80750729,80750730,80750733,80750740,80750742,80750743,80750744,80750757,80750770,80750778,80750787,80750792,80750794,80750834,80750874,80750880,80750959,80751367,80751538,80751559,80751570,80751571,80751591,80751604,80751610,80751619,80751623,80751652,80751653,80751660,80751669,80751674,80751891,80751954,80752201,80752221,80752275,80752288,80752310,80752346,80752353,80752579)),strand=c("+","+","+","+","+","+","+","+","+","+","+","+","+","+","+","+","+","+","+","+","+","+","+","+","+","+","+","+","+","+","+","+","+","+","+","+","+","+","+","+","+","+","+","+","+","+","+","+","+","+","+","+","+","+","+","+","+","+","+","+","+","+","+","+","+","+","+","+","+","+","+","+","+","-","-","-","-","-","-","-","-","-","-","-","-","-","-","-","-","-","-","-","-","-","-","-","-","-","-","-","-","-","-","-","-","-","-","-","-","-","-","-","-","-","-","-","-","-","-","-","-","-","-","-","-","-","-","-","-","-","-","-","-"))

server <- function(input, output, session) {
  dat_ext <- reactive({
    sz <- resize(dat,width = input$size)
  })
  output$plot1 <- renderCachedPlot({
    p1 <- autoplot(dat_ext(),aes(fill=strand),facets=strand ~ seqnames) + theme(legend.position = "none")
    p2 <- autoplot(dat_ext(),stat="coverage")
    tracks(Reads=p1,Coverage=p2,xlim=c(80750000, 80753000)) + ylab("")
  },cacheKeyExpr = {input$size})
}

ui <- fluidPage(
  mainPanel(
    sliderInput("size", "Read extended to:", min = 24, max = 500, value = 0, width='100%',step= 28,animate = animationOptions()),
    uiOutput("ext")
  ),
  mainPanel(
    plotOutput('plot1')
  )
)

shinyApp(ui, server)
